File: aws-globalaccelerator-endpointgroup/src/main/java/software/amazon/globalaccelerator/endpointgroup/CreateHandler.java
package software.amazon.globalaccelerator.endpointgroup;

import com.amazonaws.services.globalaccelerator.AWSGlobalAccelerator;
import com.amazonaws.services.globalaccelerator.model.CreateEndpointGroupRequest;
import com.amazonaws.services.globalaccelerator.model.EndpointConfiguration;
import com.amazonaws.services.globalaccelerator.model.EndpointDescription;
import com.amazonaws.services.globalaccelerator.model.EndpointGroup;
import lombok.val;
import software.amazon.cloudformation.proxy.*;
import software.amazon.globalaccelerator.arns.ListenerArn;

import java.util.List;
import java.util.stream.Collectors;

public class CreateHandler extends BaseHandler<CallbackContext> {

    @Override
    public ProgressEvent<ResourceModel, CallbackContext> handleRequest(
        final AmazonWebServicesClientProxy proxy,
        final ResourceHandlerRequest<ResourceModel> request,
        final CallbackContext callbackContext,
        final Logger logger) {
        logger.log(String.format("Create request [%s]", request));

        val agaClient = AcceleratorClientBuilder.getClient();
        val inferredCallbackContext = callbackContext != null ?
                callbackContext :
                CallbackContext.builder()
                        .stabilizationRetriesRemaining(HandlerCommons.NUMBER_OF_STATE_POLL_RETRIES)
                        .build();

        final ResourceModel model = request.getDesiredResourceState();
        if (model.getEndpointGroupArn() == null) {
            return CreateEndpointGroupStep(model, request, proxy, agaClient, logger);
        } else {
            return HandlerCommons.waitForSynchronizedStep(inferredCallbackContext, model, proxy, agaClient, logger);
        }
    }

    /**
     * Create an accelerator and create the correct progress continuation context
     */
    private ProgressEvent<ResourceModel, CallbackContext> CreateEndpointGroupStep(final ResourceModel model,
                                                                                  final ResourceHandlerRequest<ResourceModel> handlerRequest,
                                                                                  final AmazonWebServicesClientProxy proxy,
                                                                                  final AWSGlobalAccelerator agaClient,
                                                                                  final Logger logger) {
        val foundListener = HandlerCommons.getListener(model.getListenerArn(), proxy, agaClient, logger);
        if (foundListener == null) {
            return ProgressEvent.defaultFailureHandler(
                    new Exception(String.format("Failed to find listener with arn: [%s].  Can not create endpoint group.")),
                    HandlerErrorCode.NotFound);
        }

        // first thing get the accelerator associated to listener so we can update our model
        val listenerArn = new ListenerArn(model.getListenerArn());
        val accelerator = HandlerCommons.getAccelerator(listenerArn.getAcceleratorArn(), proxy, agaClient, logger);

        if (accelerator == null) {
            return ProgressEvent.defaultFailureHandler(
                    new Exception(String.format("Could not find accelerator for listener [%s]", model.getListenerArn())),
                    HandlerErrorCode.NotFound);
        }

        // now we can move forward and create the endpoint group and update model with everything that is optional
        val endpointGroup = CreateEndpointGroup(model, handlerRequest, proxy, agaClient);
        model.setEndpointGroupArn(endpointGroup.getEndpointGroupArn());
        model.setHealthCheckIntervalSeconds(endpointGroup.getHealthCheckIntervalSeconds());
        model.setHealthCheckPath(endpointGroup.getHealthCheckPath());
        model.setHealthCheckPort(endpointGroup.getHealthCheckPort());
        model.setHealthCheckProtocol(endpointGroup.getHealthCheckProtocol());
        model.setThresholdCount(endpointGroup.getThresholdCount());
        model.setEndpointConfigurations(getEndpointConfigurations(endpointGroup.getEndpointDescriptions()));

        val callbackContext = CallbackContext.builder()
                .stabilizationRetriesRemaining(HandlerCommons.NUMBER_OF_STATE_POLL_RETRIES)
                .build();

        return ProgressEvent.defaultInProgressHandler(callbackContext, 0, model);
    }

    private List<software.amazon.globalaccelerator.endpointgroup.EndpointConfiguration> getEndpointConfigurations (List<EndpointDescription> endpointDescriptions) {
        return endpointDescriptions.stream().map((x) -> software.amazon.globalaccelerator.endpointgroup.EndpointConfiguration.builder()
                .clientIPPreservationEnabled(x.getClientIPPreservationEnabled())
                .endpointId(x.getEndpointId())
                .weight(x.getWeight())
                .build()).collect(Collectors.toList());
    }

    private EndpointGroup CreateEndpointGroup(final ResourceModel model,
                                    final ResourceHandlerRequest<ResourceModel> handlerRequest,
                                    final AmazonWebServicesClientProxy proxy,
                                    final AWSGlobalAccelerator agaClient) {
        // If health check port is not provided we want to fall back
        Integer healthCheckPort = (model.getHealthCheckPort() < 0) ? null : model.getHealthCheckPort();

        // we need to map all of our endpoint configurations
        List<EndpointConfiguration> convertedEndpointConfigurations = null;
        if (model.getEndpointConfigurations() != null) {
            convertedEndpointConfigurations = model.getEndpointConfigurations().stream()
                    .map((x) -> new EndpointConfiguration().withEndpointId(x.getEndpointId()).withWeight(x.getWeight())
                            .withClientIPPreservationEnabled(x.getClientIPPreservationEnabled()))
                    .collect(Collectors.toList());
        }

        val createEndpointGroupRequest = new CreateEndpointGroupRequest()
                .withListenerArn(model.getListenerArn())
                .withEndpointGroupRegion(model.getEndpointGroupRegion())
                .withHealthCheckPort(healthCheckPort)
                .withHealthCheckIntervalSeconds(model.getHealthCheckIntervalSeconds())
                .withHealthCheckProtocol(model.getHealthCheckProtocol())
                .withHealthCheckPath(model.getHealthCheckPath())
                .withThresholdCount(model.getThresholdCount())
                .withEndpointConfigurations(convertedEndpointConfigurations)
                .withIdempotencyToken(handlerRequest.getLogicalResourceIdentifier());

        return proxy.injectCredentialsAndInvoke(createEndpointGroupRequest, agaClient::createEndpointGroup).getEndpointGroup();
    }
}

File: aws-globalaccelerator-endpointgroup/src/main/kotlin/software/amazon/globalaccelerator/endpointgroup/CreateHandler.kt
package software.amazon.globalaccelerator.endpointgroup

import com.amazonaws.services.globalaccelerator.AWSGlobalAccelerator
import com.amazonaws.services.globalaccelerator.model.CreateEndpointGroupRequest
import com.amazonaws.services.globalaccelerator.model.EndpointConfiguration
import com.amazonaws.services.globalaccelerator.model.EndpointGroup
import com.amazonaws.services.globalaccelerator.model.PortOverride
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy
import software.amazon.cloudformation.proxy.Logger
import software.amazon.cloudformation.proxy.ProgressEvent
import software.amazon.cloudformation.proxy.ResourceHandlerRequest
import software.amazon.cloudformation.proxy.HandlerErrorCode
import software.amazon.globalaccelerator.arns.ListenerArn

/**
 * Create handler implementation for Endpoint Group resource.
 */
class CreateHandler : BaseHandler<CallbackContext?>() {
    override fun handleRequest(proxy: AmazonWebServicesClientProxy,
                               request: ResourceHandlerRequest<ResourceModel>,
                               callbackContext: CallbackContext?,
                               logger: Logger): ProgressEvent<ResourceModel, CallbackContext?> {
        logger.debug("Create new EndpointGroup request: $request")
        val agaClient = AcceleratorClientBuilder.client
        val inferredCallbackContext = callbackContext ?: CallbackContext(stabilizationRetriesRemaining = HandlerCommons.NUMBER_OF_STATE_POLL_RETRIES)
        val model = request.desiredResourceState
        return when (model.endpointGroupArn) {
            null -> createEndpointGroupStep(model, request, proxy, agaClient, logger)
            else -> HandlerCommons.waitForSynchronizedStep(inferredCallbackContext, model, proxy, agaClient, logger)
        }
    }

    private fun createEndpointGroupStep(model: ResourceModel,
                                        handlerRequest: ResourceHandlerRequest<ResourceModel>,
                                        proxy: AmazonWebServicesClientProxy,
                                        agaClient: AWSGlobalAccelerator,
                                        logger: Logger): ProgressEvent<ResourceModel, CallbackContext?> {
        HandlerCommons.getListener(model.listenerArn, proxy, agaClient, logger)
                ?: return ProgressEvent.defaultFailureHandler(
                        Exception("Failed to create Endpoint Group. Cannot find listener with arn: [${model.listenerArn}]."),
                        HandlerErrorCode.NotFound)

        val listenerArn = ListenerArn(model.listenerArn)
        // Make sure accelerator exists
        HandlerCommons.getAccelerator(listenerArn.acceleratorArn, proxy, agaClient, logger)
                ?: return ProgressEvent.defaultFailureHandler(
                        Exception("Could not find accelerator for listener with arn: [${model.listenerArn}]."),
                        HandlerErrorCode.NotFound)
        val endpointGroup = createEndpointGroup(model, handlerRequest, proxy, agaClient)
        model.apply {
            this.endpointGroupArn = endpointGroup.endpointGroupArn
            this.healthCheckIntervalSeconds = endpointGroup.healthCheckIntervalSeconds
            this.healthCheckPath = endpointGroup.healthCheckPath
            this.healthCheckPort = endpointGroup.healthCheckPort
            this.healthCheckProtocol = endpointGroup.healthCheckProtocol
            this.thresholdCount = endpointGroup.thresholdCount
            this.trafficDialPercentage = endpointGroup.trafficDialPercentage.toDouble()
            this.endpointConfigurations = getEndpointConfigurations(endpointGroup.endpointDescriptions)
            this.portOverrides = getPortOverrides(endpointGroup.portOverrides)
        }
        val callbackContext = CallbackContext(stabilizationRetriesRemaining = HandlerCommons.NUMBER_OF_STATE_POLL_RETRIES)
        return ProgressEvent.defaultInProgressHandler(callbackContext, 0, model)
    }

    private fun createEndpointGroup(model: ResourceModel,
                                    handlerRequest: ResourceHandlerRequest<ResourceModel>,
                                    proxy: AmazonWebServicesClientProxy,
                                    agaClient: AWSGlobalAccelerator): EndpointGroup {
        // Mapping all endpoint configurations
        val convertedEndpointConfigurations = model.endpointConfigurations?.map {EndpointConfiguration()
                    .withEndpointId(it.endpointId).withWeight(it.weight).withAttachmentArn(it.attachmentArn)
                    .withClientIPPreservationEnabled(it.clientIPPreservationEnabled)}
        val trafficDialPercentage = model.trafficDialPercentage?.toFloat() ?: 100.0f
        val portOverrides = model.portOverrides?.map { PortOverride().withListenerPort(it.listenerPort).withEndpointPort(it.endpointPort) }
        val createEndpointGroupRequest = CreateEndpointGroupRequest()
                .withListenerArn(model.listenerArn)
                .withEndpointGroupRegion(model.endpointGroupRegion)
                .withHealthCheckPort(model.healthCheckPort)
                .withHealthCheckIntervalSeconds(model.healthCheckIntervalSeconds)
                .withHealthCheckProtocol(model.healthCheckProtocol)
                .withHealthCheckPath(model.healthCheckPath)
                .withThresholdCount(model.thresholdCount)
                .withTrafficDialPercentage(trafficDialPercentage)
                .withEndpointConfigurations(convertedEndpointConfigurations)
                .withPortOverrides(portOverrides)
                .withIdempotencyToken(handlerRequest.clientRequestToken)
        return proxy.injectCredentialsAndInvoke(createEndpointGroupRequest, agaClient::createEndpointGroup).endpointGroup
    }
}
